File: alms_app/crms/web_form/car_allowance/car_allowance.js
console.log("✅ car_allowance_form.js loaded");

frappe.ready(function () {
    function getEmployeeCodeFromURL() {
        const params = new URLSearchParams(window.location.search);
        return decodeURIComponent(params.get("employee_code") || "");
    }

    const employeeCode = getEmployeeCodeFromURL();
    if (employeeCode) {
        frappe.web_form.set_value('employee_code', employeeCode);

        frappe.call({
            method: "alms_app.crms.web_form.car_allowance.car_allowance.get_employee_details",
            args: { employee_code: employeeCode },
            callback: function (response) {
                const employeeDetails = response.message;
                if (employeeDetails) {
                    frappe.web_form.set_value('designation', employeeDetails[0].designation);
                    frappe.web_form.set_value('location', employeeDetails[0].location);
                    frappe.web_form.set_value('company_name', employeeDetails[0].company);
                    frappe.web_form.set_value('employee_reporting', employeeDetails[0].reporting_head);
                    frappe.web_form.set_value('contact_number', employeeDetails[0].contact_number);
                    frappe.web_form.set_value('email_id', employeeDetails[0].email_id);
                    frappe.web_form.set_value('department', employeeDetails[0].department);
                    frappe.web_form.set_value('eligibility', employeeDetails[0].eligibility);
                }
            }
        });
    }

    frappe.web_form.validate = function () {
        const employeeCode = frappe.web_form.get_value("employee_code");
        let exists = false;
        
        // Synchronous check before Frappe's native save
        frappe.call({
            method: "alms_app.crms.web_form.car_allowance.car_allowance.check_allowance_exists",
            args: { employee_code: employeeCode },
            async: false,
            callback: function (r) {
                if (r.message === "exists") {
                    exists = true;
                }
            }
        });

        if (exists) {
            frappe.show_alert({
                message: __('Allowance already exists.'),
                indicator: 'red'
            }, 3);
            setTimeout(() => { window.location.href = "/already-present"; }, 2000);
            return false; // Aborts native save
        }
        return true; // Proceeds to native save -> after_save
    };

    frappe.web_form.after_save = function () {
        const employeeCode = frappe.web_form.get_value("employee_code");
        frappe.call({
            method: "alms_app.crms.web_form.car_allowance.car_allowance.send_allowance_email",
            args: { employee_code: employeeCode },
            callback: function () {
                frappe.show_alert({
                    message: __('Your Car Allowance request was submitted successfully.'),
                    indicator: 'green'
                }, 3);
                setTimeout(() => { window.location.href = "/allowance"; }, 2000);
            }
        });
    };
});

File: alms_app/crms/web_form/car_allowance/car_allowance.py
import frappe
from frappe.core.doctype.communication.email import make
from email.mime.text import MIMEText
import smtplib
from alms_app.newutils.custom_sendmail import custom_sendmail
from alms_app.api.emailsService import email_sender


@frappe.whitelist(allow_guest=True)
def get_employee_details(employee_code):
    employee = frappe.get_all("Employee", filters={"name": employee_code}, fields=["*"])
    
    if not employee:
        frappe.throw(f"Employee with name '{employee_code}' not found.")
        
    emp = employee[0]
        
    # Determine eligibility from Employee Designation or fallback to inline value
    eligibility_val = emp.get("eligibility") or 0
    if emp.get("meril_designation"):
        designation_eligibility = frappe.db.get_value("Employee Designation", emp.meril_designation, "eligibility")
        if designation_eligibility is not None:
            eligibility_val = designation_eligibility
            
    # Map fields for frontend compatibility
    emp["eligibility"] = eligibility_val
    emp["location"] = emp.get("branch")
    emp["contact_number"] = emp.get("cell_number")
    emp["email_id"] = emp.get("company_email") or emp.get("user_id")
    
    return [emp]

    
def get_context(context):
    context.employee_code = frappe.form_dict.get('employee_code')
    context.employee_details = {}

    if context.employee_code:
        employee_details = get_employee_details(context.employee_code)
        context.employee_details = employee_details

    return context

    
@frappe.whitelist(allow_guest=True)
def send_allowance_email(employee_code):
    from alms_app.api.emailClass import EmailServices
    return EmailServices().send_allowance_email(employee_code)
    
@frappe.whitelist(allow_guest=True)
def check_allowance_exists(employee_code):
    exists = frappe.db.exists("Car Allowance", {"employee_code": employee_code, "docstatus": ["<", 2]})
    return "exists" if exists else "not_exists"

@frappe.whitelist(allow_guest=True)
def create_allowance_entry(employee_code):
    # Check again to avoid duplicates if called directly
    exists = frappe.db.exists("Car Allowance", {"employee_code": employee_code, "docstatus": ["<", 2]})
    if exists:
        return {"status": "failed", "message": "Allowance already exists"}

    try:
        # Create new entry
        doc = frappe.new_doc("Car Allowance")
        doc.employee_code = employee_code
        doc.insert(ignore_permissions=True)
        frappe.db.commit()
    except frappe.exceptions.DuplicateEntryError:
        return {"status": "failed", "message": f"Car Allowance {employee_code} already exists"}

    # Trigger email via your existing method
    send_allowance_email(employee_code)

    return {"status": "success", "message": f"Allowance created for {employee_code}"}
